File: src/plugins/api-provider.ts
/**
 * 플러그인 API 프로바이더
 * 플러그인이 사용할 수 있는 API 컨텍스트 제공
 */

import { EventEmitter } from 'events';
import { promises as fs } from 'fs';
import axios, { AxiosInstance } from 'axios';
import {
  PluginAPIContext,
  PluginDescriptor,
  PluginLogger,
  PluginFileSystem,
  PluginHTTPClient,
  PluginDatabase,
  PluginMCPTools,
  PluginNotifications,
  PluginCommunication,
  PluginStorage,
  PluginPermission
} from './types.js';

/**
 * 플러그인 API 프로바이더
 */
export class PluginAPIProvider {
  private contexts = new Map<string, PluginAPIContext>();
  private eventBus = new EventEmitter();
  private storages = new Map<string, Map<string, any>>();

  constructor() {
    this.eventBus.setMaxListeners(1000); // 많은 플러그인을 위해 리스너 한도 증가
  }

  /**
   * 플러그인용 API 컨텍스트 생성
   */
  async createContext(descriptor: PluginDescriptor): Promise<PluginAPIContext> {
    const pluginId = descriptor.id;
    const permissions = descriptor.manifest.permissions || [];

    const context: PluginAPIContext = {
      metadata: descriptor.manifest,
      config: descriptor.manifest.configSchema || {},
      events: this.createEventInterface(pluginId, permissions),
      logger: this.createLogger(pluginId),
      fs: this.createFileSystemInterface(pluginId, permissions),
      http: this.createHTTPInterface(pluginId, permissions),
      database: this.createDatabaseInterface(pluginId, permissions),
      mcp: this.createMCPInterface(pluginId, permissions),
      notifications: this.createNotificationInterface(pluginId, permissions),
      communication: this.createCommunicationInterface(pluginId),
      storage: this.createStorageInterface(pluginId)
    };

    this.contexts.set(pluginId, context);
    return context;
  }

  /**
   * API 컨텍스트 해제
   */
  async destroyContext(pluginId: string): Promise<void> {
    const context = this.contexts.get(pluginId);
    if (!context) {
      return;
    }

    // 이벤트 리스너 정리
    this.eventBus.removeAllListeners(`plugin.${pluginId}.*`);
    
    // 스토리지 정리
    this.storages.delete(pluginId);
    
    // 컨텍스트 제거
    this.contexts.delete(pluginId);
  }

  /**
   * 이벤트 인터페이스 생성
   */
  private createEventInterface(pluginId: string, permissions: PluginPermission[]): EventEmitter {
    const eventInterface = new EventEmitter();
    
    // 이벤트 읽기 권한 체크
    const canReadEvents = permissions.includes(PluginPermission.READ_EVENTS);
    const canWriteEvents = permissions.includes(PluginPermission.WRITE_EVENTS);

    if (canReadEvents) {
      // 글로벌 이벤트를 플러그인에 전달
      this.eventBus.on('*', (event, data) => {
        eventInterface.emit(event, data);
      });
    }

    if (canWriteEvents) {
      // 플러그인 이벤트를 글로벌 버스에 전달
      eventInterface.on('*', (event, data) => {
        this.eventBus.emit(`plugin.${pluginId}.${event}`, data);
        this.eventBus.emit(event, { ...data, source: pluginId });
      });
    }

    return eventInterface;
  }

  /**
   * 로거 인터페이스 생성
   */
  private createLogger(pluginId: string): PluginLogger {
    const prefix = `[Plugin:${pluginId}]`;
    
    return {
      debug: (message: string, meta?: any) => {
        console.debug(`${prefix} [DEBUG] ${message}`, meta || '');
      },
      
      info: (message: string, meta?: any) => {
        console.info(`${prefix} [INFO] ${message}`, meta || '');
      },
      
      warn: (message: string, meta?: any) => {
        console.warn(`${prefix} [WARN] ${message}`, meta || '');
      },
      
      error: (message: string, error?: Error, meta?: any) => {
        console.error(`${prefix} [ERROR] ${message}`, error || '', meta || '');
      }
    };
  }

  /**
   * 파일 시스템 인터페이스 생성
   */
  private createFileSystemInterface(_pluginId: string, permissions: PluginPermission[]): PluginFileSystem {
    const canReadFiles = permissions.includes(PluginPermission.READ_FILES);
    const canWriteFiles = permissions.includes(PluginPermission.WRITE_FILES);

    return {
      readFile: async (path: string): Promise<string> => {
        if (!canReadFiles) {
          throw new Error('Plugin does not have file read permission');
        }
        
        // 보안: 플러그인 디렉토리 외부 접근 제한
        if (path.includes('..') || !path.startsWith('/')) {
          throw new Error('Invalid file path');
        }
        
        return fs.readFile(path, 'utf-8');
      },

      writeFile: async (path: string, content: string): Promise<void> => {
        if (!canWriteFiles) {
          throw new Error('Plugin does not have file write permission');
        }
        
        // 보안: 플러그인 디렉토리 외부 접근 제한
        if (path.includes('..') || !path.startsWith('/')) {
          throw new Error('Invalid file path');
        }
        
        await fs.writeFile(path, content, 'utf-8');
      },

      exists: async (path: string): Promise<boolean> => {
        if (!canReadFiles) {
          return false;
        }
        
        try {
          await fs.access(path);
          return true;
        } catch {
          return false;
        }
      },

      mkdir: async (path: string): Promise<void> => {
        if (!canWriteFiles) {
          throw new Error('Plugin does not have file write permission');
        }
        
        await fs.mkdir(path, { recursive: true });
      },

      readDir: async (path: string): Promise<string[]> => {
        if (!canReadFiles) {
          throw new Error('Plugin does not have file read permission');
        }
        
        const entries = await fs.readdir(path);
        return entries;
      },

      watch: (path: string, callback: (event: string, filename: string) => void): void => {
        if (!canReadFiles) {
          throw new Error('Plugin does not have file read permission');
        }
        
        // 파일 감시는 chokidar나 fs.watch 사용
        const { watch } = require('fs');
        watch(path, { recursive: true }, callback);
      }
    };
  }

  /**
   * HTTP 클라이언트 인터페이스 생성
   */
  private createHTTPInterface(pluginId: string, permissions: PluginPermission[]): PluginHTTPClient {
    const canAccessNetwork = permissions.includes(PluginPermission.NETWORK_ACCESS);
    
    if (!canAccessNetwork) {
      // 네트워크 접근 권한이 없는 경우 더미 인터페이스 반환
      const noPermissionError = () => {
        throw new Error('Plugin does not have network access permission');
      };
      
      return {
        get: noPermissionError,
        post: noPermissionError,
        put: noPermissionError,
        delete: noPermissionError
      };
    }

    const httpClient: AxiosInstance = axios.create({
      timeout: 30000,
      headers: {
        'User-Agent': `DevFlowMonitor-Plugin/${pluginId}`
      }
    });

    return {
      get: async (url: string, options?: any) => {
        const response = await httpClient.get(url, options);
        return response.data;
      },

      post: async (url: string, data: any, options?: any) => {
        const response = await httpClient.post(url, data, options);
        return response.data;
      },

      put: async (url: string, data: any, options?: any) => {
        const response = await httpClient.put(url, data, options);
        return response.data;
      },

      delete: async (url: string, options?: any) => {
        const response = await httpClient.delete(url, options);
        return response.data;
      }
    };
  }

  /**
   * 데이터베이스 인터페이스 생성
   */
  private createDatabaseInterface(_pluginId: string, permissions: PluginPermission[]): PluginDatabase {
    const canReadDB = permissions.includes(PluginPermission.DATABASE_READ);
    const canWriteDB = permissions.includes(PluginPermission.DATABASE_WRITE);

    return {
      query: async (sql: string, _params?: any[]): Promise<any[]> => {
        if (!canReadDB && !sql.toLowerCase().trim().startsWith('select')) {
          throw new Error('Plugin does not have database read permission');
        }
        if (!canWriteDB && !sql.toLowerCase().trim().startsWith('select')) {
          throw new Error('Plugin does not have database write permission');
        }
        
        // 실제 데이터베이스 쿼리는 StorageManager를 통해 실행
        // 여기서는 플레이스홀더
        throw new Error('Database interface not implemented yet');
      },

      insert: async (_table: string, _data: Record<string, any>): Promise<number> => {
        if (!canWriteDB) {
          throw new Error('Plugin does not have database write permission');
        }
        
        // 실제 구현 필요
        throw new Error('Database interface not implemented yet');
      },

      update: async (_table: string, _data: Record<string, any>, _where: Record<string, any>): Promise<number> => {
        if (!canWriteDB) {
          throw new Error('Plugin does not have database write permission');
        }
        
        // 실제 구현 필요
        throw new Error('Database interface not implemented yet');
      },

      delete: async (_table: string, _where: Record<string, any>): Promise<number> => {
        if (!canWriteDB) {
          throw new Error('Plugin does not have database write permission');
        }
        
        // 실제 구현 필요
        throw new Error('Database interface not implemented yet');
      }
    };
  }

  /**
   * MCP 도구 인터페이스 생성
   */
  private createMCPInterface(pluginId: string, permissions: PluginPermission[]): PluginMCPTools {
    const canUseMCPTools = permissions.includes(PluginPermission.MCP_TOOLS);
    
    const registeredTools = new Map<string, Function>();

    return {
      registerTool: (name: string, _description: string, handler: Function): void => {
        if (!canUseMCPTools) {
          throw new Error('Plugin does not have MCP tools permission');
        }
        
        const toolName = `plugin_${pluginId}_${name}`;
        registeredTools.set(toolName, handler);
        
        // 실제 MCP 서버에 도구 등록 (나중에 구현)
        console.log(`[PluginAPI] Tool registered: ${toolName}`);
      },

      unregisterTool: (name: string): void => {
        if (!canUseMCPTools) {
          throw new Error('Plugin does not have MCP tools permission');
        }
        
        const toolName = `plugin_${pluginId}_${name}`;
        registeredTools.delete(toolName);
        
        // 실제 MCP 서버에서 도구 제거 (나중에 구현)
        console.log(`[PluginAPI] Tool unregistered: ${toolName}`);
      },

      callTool: async (name: string, args: any): Promise<any> => {
        if (!canUseMCPTools) {
          throw new Error('Plugin does not have MCP tools permission');
        }
        
        const handler = registeredTools.get(name);
        if (!handler) {
          throw new Error(`Tool not found: ${name}`);
        }
        
        return handler(args);
      }
    };
  }

  /**
   * 알림 인터페이스 생성
   */
  private createNotificationInterface(pluginId: string, permissions: PluginPermission[]): PluginNotifications {
    const canSendNotifications = permissions.includes(PluginPermission.NOTIFICATIONS);

    return {
      send: async (message: string, level: 'info' | 'warning' | 'error', options?: any): Promise<void> => {
        if (!canSendNotifications) {
          throw new Error('Plugin does not have notification permission');
        }
        
        // 실제 알림 시스템에 전달 (나중에 구현)
        console.log(`[Plugin:${pluginId}] Notification [${level.toUpperCase()}]: ${message}`);
        
        this.eventBus.emit('plugin.notification', {
          pluginId,
          message,
          level,
          options,
          timestamp: new Date()
        });
      },

      createRule: async (_rule: any): Promise<string> => {
        if (!canSendNotifications) {
          throw new Error('Plugin does not have notification permission');
        }
        
        // 알림 규칙 생성 (나중에 구현)
        const ruleId = `plugin_${pluginId}_${Date.now()}`;
        console.log(`[Plugin:${pluginId}] Notification rule created: ${ruleId}`);
        return ruleId;
      },

      removeRule: async (ruleId: string): Promise<void> => {
        if (!canSendNotifications) {
          throw new Error('Plugin does not have notification permission');
        }
        
        // 알림 규칙 제거 (나중에 구현)
        console.log(`[Plugin:${pluginId}] Notification rule removed: ${ruleId}`);
      }
    };
  }

  /**
   * 플러그인 간 통신 인터페이스 생성
   */
  private createCommunicationInterface(pluginId: string): PluginCommunication {
    return {
      sendMessage: async (targetPlugin: string, message: any): Promise<void> => {
        this.eventBus.emit(`plugin.message.${targetPlugin}`, {
          from: pluginId,
          to: targetPlugin,
          message,
          timestamp: new Date()
        });
      },

      broadcast: (event: string, data: any): void => {
        this.eventBus.emit(`plugin.broadcast.${event}`, {
          from: pluginId,
          event,
          data,
          timestamp: new Date()
        });
      },

      subscribe: (event: string, handler: (...args: any[]) => void): void => {
        this.eventBus.on(`plugin.broadcast.${event}`, handler);
        this.eventBus.on(`plugin.message.${pluginId}`, handler);
      },

      unsubscribe: (event: string, handler: (...args: any[]) => void): void => {
        this.eventBus.off(`plugin.broadcast.${event}`, handler);
        this.eventBus.off(`plugin.message.${pluginId}`, handler);
      }
    };
  }

  /**
   * 스토리지 인터페이스 생성
   */
  private createStorageInterface(pluginId: string): PluginStorage {
    // 플러그인별 독립적인 스토리지 공간
    if (!this.storages.has(pluginId)) {
      this.storages.set(pluginId, new Map<string, any>());
    }
    
    const storage = this.storages.get(pluginId)!;

    return {
      get: async (key: string): Promise<any> => {
        return storage.get(key);
      },

      set: async (key: string, value: any): Promise<void> => {
        storage.set(key, value);
        
        // 지속성을 위해 파일에 저장 (나중에 구현)
        // await this.persistPluginStorage(pluginId, storage);
      },

      delete: async (key: string): Promise<void> => {
        storage.delete(key);
        
        // 지속성을 위해 파일에 저장 (나중에 구현)
        // await this.persistPluginStorage(pluginId, storage);
      },

      clear: async (): Promise<void> => {
        storage.clear();
        
        // 지속성을 위해 파일에 저장 (나중에 구현)
        // await this.persistPluginStorage(pluginId, storage);
      },

      keys: async (): Promise<string[]> => {
        return Array.from(storage.keys());
      }
    };
  }

  /**
   * 글로벌 이벤트 발생
   */
  emitGlobalEvent(event: string, data: any): void {
    this.eventBus.emit(event, data);
  }

  /**
   * 글로벌 이벤트 구독
   */
  onGlobalEvent(event: string, handler: (...args: any[]) => void): void {
    this.eventBus.on(event, handler);
  }

  /**
   * 정리
   */
  async dispose(): Promise<void> {
    // 모든 이벤트 리스너 제거
    this.eventBus.removeAllListeners();
    
    // 모든 컨텍스트 정리
    this.contexts.clear();
    
    // 모든 스토리지 정리
    this.storages.clear();
    
    console.log('[PluginAPIProvider] Disposed');
  }

  /**
   * 플러그인 스토리지 지속성 (향후 구현)
   */
  // private async persistPluginStorage(_pluginId: string, _storage: Map<string, any>): Promise<void> {
  //   // JSON 파일로 저장하거나 데이터베이스에 저장
  //   // 구현 필요
  // }

  /**
   * 플러그인 스토리지 복원 (향후 구현)
   */
  // private async restorePluginStorage(_pluginId: string): Promise<Map<string, any>> {
  //   // JSON 파일에서 복원하거나 데이터베이스에서 복원
  //   // 구현 필요
  //   return new Map<string, any>();
  // }
}

File: python/src/devflow_monitor/plugins/api_provider.py
"""
Plugin API Provider.

This module provides API contexts for plugins with permission-based
access control. Each plugin receives an isolated context with only
the APIs it has permission to use.
"""

import asyncio
import os
from datetime import datetime
from typing import Any, Callable, Optional

import aiofiles
import httpx

from .types import (
    PluginAPIContext,
    PluginCommunication,
    PluginDatabase,
    PluginDescriptor,
    PluginFileSystem,
    PluginHTTPClient,
    PluginLogger,
    PluginMCPTools,
    PluginNotifications,
    PluginPermission,
    PluginStorage,
    EventHandler,
)


class PluginLoggerImpl(PluginLogger):
    """Plugin logger implementation."""

    def __init__(self, plugin_id: str):
        """Initialize logger with plugin ID prefix."""
        self.plugin_id = plugin_id
        self._prefix = f"[Plugin:{plugin_id}]"

    def debug(self, message: str, meta: Optional[Any] = None) -> None:
        """Log debug message."""
        print(f"{self._prefix} [DEBUG] {message}", meta or "")

    def info(self, message: str, meta: Optional[Any] = None) -> None:
        """Log info message."""
        print(f"{self._prefix} [INFO] {message}", meta or "")

    def warn(self, message: str, meta: Optional[Any] = None) -> None:
        """Log warning message."""
        print(f"{self._prefix} [WARN] {message}", meta or "")

    def error(
        self, message: str, error: Optional[Exception] = None, meta: Optional[Any] = None
    ) -> None:
        """Log error message."""
        print(f"{self._prefix} [ERROR] {message}", error or "", meta or "")


class PluginFileSystemImpl(PluginFileSystem):
    """Plugin file system implementation with permission checks."""

    def __init__(
        self,
        plugin_id: str,
        can_read: bool,
        can_write: bool,
        base_path: Optional[str] = None,
    ):
        """Initialize file system with permissions."""
        self.plugin_id = plugin_id
        self.can_read = can_read
        self.can_write = can_write
        self.base_path = base_path

    def _validate_path(self, path: str) -> None:
        """Validate path is safe."""
        # Prevent path traversal
        if ".." in path:
            raise ValueError("Path traversal not allowed")

        # If base path is set, ensure path is within it
        if self.base_path:
            abs_path = os.path.abspath(path)
            abs_base = os.path.abspath(self.base_path)
            if not abs_path.startswith(abs_base):
                raise ValueError("Access outside plugin directory not allowed")

    async def read_file(self, path: str) -> str:
        """Read file contents."""
        if not self.can_read:
            raise PermissionError("Plugin does not have file read permission")
        self._validate_path(path)

        async with aiofiles.open(path, "r", encoding="utf-8") as f:
            return await f.read()

    async def write_file(self, path: str, content: str) -> None:
        """Write file contents."""
        if not self.can_write:
            raise PermissionError("Plugin does not have file write permission")
        self._validate_path(path)

        async with aiofiles.open(path, "w", encoding="utf-8") as f:
            await f.write(content)

    async def exists(self, path: str) -> bool:
        """Check if path exists."""
        if not self.can_read:
            return False
        try:
            self._validate_path(path)
            return os.path.exists(path)
        except ValueError:
            return False

    async def mkdir(self, path: str) -> None:
        """Create directory."""
        if not self.can_write:
            raise PermissionError("Plugin does not have file write permission")
        self._validate_path(path)
        os.makedirs(path, exist_ok=True)

    async def read_dir(self, path: str) -> list[str]:
        """Read directory contents."""
        if not self.can_read:
            raise PermissionError("Plugin does not have file read permission")
        self._validate_path(path)
        return os.listdir(path)

    def watch(self, path: str, callback: Callable[[str, str], None]) -> None:
        """Watch path for changes."""
        if not self.can_read:
            raise PermissionError("Plugin does not have file read permission")
        self._validate_path(path)
        # Would integrate with watchdog here
        raise NotImplementedError("File watching not yet implemented")


class PluginHTTPClientImpl(PluginHTTPClient):
    """Plugin HTTP client implementation with permission checks."""

    def __init__(self, plugin_id: str, allowed: bool, timeout: float = 30.0):
        """Initialize HTTP client with permissions."""
        self.plugin_id = plugin_id
        self.allowed = allowed
        self.timeout = timeout
        self._client: Optional[httpx.AsyncClient] = None

    async def _get_client(self) -> httpx.AsyncClient:
        """Get or create HTTP client."""
        if not self._client:
            self._client = httpx.AsyncClient(
                timeout=self.timeout,
                headers={"User-Agent": f"DevFlowMonitor-Plugin/{self.plugin_id}"},
            )
        return self._client

    def _check_permission(self) -> None:
        """Check if network access is allowed."""
        if not self.allowed:
            raise PermissionError("Plugin does not have network access permission")

    async def get(self, url: str, options: Optional[dict[str, Any]] = None) -> Any:
        """Make GET request."""
        self._check_permission()
        client = await self._get_client()
        response = await client.get(url, **(options or {}))
        response.raise_for_status()
        return response.json()

    async def post(
        self, url: str, data: Any, options: Optional[dict[str, Any]] = None
    ) -> Any:
        """Make POST request."""
        self._check_permission()
        client = await self._get_client()
        response = await client.post(url, json=data, **(options or {}))
        response.raise_for_status()
        return response.json()

    async def put(
        self, url: str, data: Any, options: Optional[dict[str, Any]] = None
    ) -> Any:
        """Make PUT request."""
        self._check_permission()
        client = await self._get_client()
        response = await client.put(url, json=data, **(options or {}))
        response.raise_for_status()
        return response.json()

    async def delete(self, url: str, options: Optional[dict[str, Any]] = None) -> Any:
        """Make DELETE request."""
        self._check_permission()
        client = await self._get_client()
        response = await client.delete(url, **(options or {}))
        response.raise_for_status()
        return response.json()

    async def close(self) -> None:
        """Close the HTTP client."""
        if self._client:
            await self._client.aclose()
            self._client = None


class PluginDatabaseImpl(PluginDatabase):
    """Plugin database implementation with permission checks."""

    def __init__(self, plugin_id: str, can_read: bool, can_write: bool):
        """Initialize database with permissions."""
        self.plugin_id = plugin_id
        self.can_read = can_read
        self.can_write = can_write

    async def query(self, sql: str, params: Optional[list[Any]] = None) -> list[Any]:
        """Execute SQL query."""
        sql_lower = sql.strip().lower()
        is_select = sql_lower.startswith("select")

        if not is_select and not self.can_write:
            raise PermissionError("Plugin does not have database write permission")
        if is_select and not self.can_read:
            raise PermissionError("Plugin does not have database read permission")

        # Would integrate with actual database here
        raise NotImplementedError("Database interface not implemented yet")

    async def insert(self, table: str, data: dict[str, Any]) -> int:
        """Insert record."""
        if not self.can_write:
            raise PermissionError("Plugin does not have database write permission")
        raise NotImplementedError("Database interface not implemented yet")

    async def update(
        self, table: str, data: dict[str, Any], where: dict[str, Any]
    ) -> int:
        """Update records."""
        if not self.can_write:
            raise PermissionError("Plugin does not have database write permission")
        raise NotImplementedError("Database interface not implemented yet")

    async def delete_record(self, table: str, where: dict[str, Any]) -> int:
        """Delete records."""
        if not self.can_write:
            raise PermissionError("Plugin does not have database write permission")
        raise NotImplementedError("Database interface not implemented yet")


class PluginMCPToolsImpl(PluginMCPTools):
    """Plugin MCP tools implementation with permission checks."""

    def __init__(self, plugin_id: str, allowed: bool):
        """Initialize MCP tools with permissions."""
        self.plugin_id = plugin_id
        self.allowed = allowed
        self._registered_tools: dict[str, Callable[..., Any]] = {}

    def _check_permission(self) -> None:
        """Check if MCP tools are allowed."""
        if not self.allowed:
            raise PermissionError("Plugin does not have MCP tools permission")

    def register_tool(
        self, name: str, description: str, handler: Callable[..., Any]
    ) -> None:
        """Register an MCP tool."""
        self._check_permission()
        tool_name = f"plugin_{self.plugin_id}_{name}"
        self._registered_tools[tool_name] = handler

    def unregister_tool(self, name: str) -> None:
        """Unregister an MCP tool."""
        self._check_permission()
        tool_name = f"plugin_{self.plugin_id}_{name}"
        if tool_name in self._registered_tools:
            del self._registered_tools[tool_name]

    async def call_tool(self, name: str, args: Any) -> Any:
        """Call an MCP tool."""
        self._check_permission()
        handler = self._registered_tools.get(name)
        if not handler:
            raise ValueError(f"Tool not found: {name}")

        result = handler(args)
        if asyncio.iscoroutine(result):
            return await result
        return result


class PluginNotificationsImpl(PluginNotifications):
    """Plugin notifications implementation with permission checks."""

    def __init__(self, plugin_id: str, allowed: bool, event_bus: Any):
        """Initialize notifications with permissions."""
        self.plugin_id = plugin_id
        self.allowed = allowed
        self.event_bus = event_bus

    def _check_permission(self) -> None:
        """Check if notifications are allowed."""
        if not self.allowed:
            raise PermissionError("Plugin does not have notification permission")

    async def send(
        self,
        message: str,
        level: str = "info",
        options: Optional[dict[str, Any]] = None,
    ) -> None:
        """Send notification."""
        self._check_permission()
        self.event_bus.emit("plugin.notification", {
            "plugin_id": self.plugin_id,
            "message": message,
            "level": level,
            "options": options,
            "timestamp": datetime.utcnow(),
        })

    async def create_rule(self, rule: Any) -> str:
        """Create notification rule."""
        self._check_permission()
        rule_id = f"plugin_{self.plugin_id}_{datetime.utcnow().timestamp()}"
        return rule_id

    async def remove_rule(self, rule_id: str) -> None:
        """Remove notification rule."""
        self._check_permission()


class PluginCommunicationImpl(PluginCommunication):
    """Plugin inter-plugin communication implementation."""

    def __init__(self, plugin_id: str, event_bus: Any):
        """Initialize communication with event bus."""
        self.plugin_id = plugin_id
        self.event_bus = event_bus

    async def send_message(self, target_plugin: str, message: Any) -> None:
        """Send message to another plugin."""
        self.event_bus.emit(f"plugin.message.{target_plugin}", {
            "from": self.plugin_id,
            "to": target_plugin,
            "message": message,
            "timestamp": datetime.utcnow(),
        })

    def broadcast(self, event: str, data: Any) -> None:
        """Broadcast event to all plugins."""
        self.event_bus.emit(f"plugin.broadcast.{event}", {
            "from": self.plugin_id,
            "event": event,
            "data": data,
            "timestamp": datetime.utcnow(),
        })

    def subscribe(self, event: str, handler: EventHandler) -> None:
        """Subscribe to event."""
        self.event_bus.on(f"plugin.broadcast.{event}", handler)
        self.event_bus.on(f"plugin.message.{self.plugin_id}", handler)

    def unsubscribe(self, event: str, handler: EventHandler) -> None:
        """Unsubscribe from event."""
        # Would need event bus with off() method
        pass


class PluginStorageImpl(PluginStorage):
    """Plugin storage implementation with isolated namespace."""

    def __init__(self, plugin_id: str):
        """Initialize storage with plugin namespace."""
        self.plugin_id = plugin_id
        self._data: dict[str, Any] = {}

    async def get(self, key: str) -> Any:
        """Get value by key."""
        return self._data.get(key)

    async def set(self, key: str, value: Any) -> None:
        """Set value by key."""
        self._data[key] = value

    async def delete(self, key: str) -> None:
        """Delete value by key."""
        if key in self._data:
            del self._data[key]

    async def clear(self) -> None:
        """Clear all storage."""
        self._data.clear()

    async def keys(self) -> list[str]:
        """Get all keys."""
        return list(self._data.keys())


class EventBus:
    """Simple event bus for plugin communication."""

    def __init__(self):
        """Initialize event bus."""
        self._handlers: dict[str, list[Callable[..., Any]]] = {}
        self._max_listeners = 1000

    def on(self, event: str, handler: Callable[..., Any]) -> None:
        """Register event handler."""
        if event not in self._handlers:
            self._handlers[event] = []
        if len(self._handlers[event]) < self._max_listeners:
            self._handlers[event].append(handler)

    def off(self, event: str, handler: Callable[..., Any]) -> None:
        """Remove event handler."""
        if event in self._handlers:
            try:
                self._handlers[event].remove(handler)
            except ValueError:
                pass

    def emit(self, event: str, data: Any) -> None:
        """Emit event to all handlers."""
        handlers = self._handlers.get(event, [])
        for handler in handlers:
            try:
                result = handler(data)
                if asyncio.iscoroutine(result):
                    asyncio.create_task(result)
            except Exception:
                pass

    def remove_all(self, prefix: str) -> None:
        """Remove all handlers with given prefix."""
        to_remove = [k for k in self._handlers.keys() if k.startswith(prefix)]
        for key in to_remove:
            del self._handlers[key]


class PluginAPIProvider:
    """
    Plugin API Provider.

    Creates isolated API contexts for plugins with permission-based
    access control. Each plugin receives only the APIs it has permission to use.

    Example:
        >>> provider = PluginAPIProvider()
        >>> context = await provider.create_context(descriptor)
        >>> # context now has logger, fs, http, etc. based on permissions
    """

    def __init__(self):
        """Initialize the API provider."""
        self._contexts: dict[str, PluginAPIContext] = {}
        self._storages: dict[str, PluginStorageImpl] = {}
        self._http_clients: dict[str, PluginHTTPClientImpl] = {}
        self._event_bus = EventBus()

    async def create_context(self, descriptor: PluginDescriptor) -> PluginAPIContext:
        """
        Create an API context for a plugin.

        Args:
            descriptor: Plugin descriptor.

        Returns:
            PluginAPIContext with all available APIs.
        """
        plugin_id = descriptor.id
        permissions = descriptor.manifest.permissions or []
        perm_set = set(permissions)

        # Create base context
        context = PluginAPIContext(
            metadata=descriptor.manifest,
            config=descriptor.manifest.config_schema or {},
        )

        # Create logger (always available)
        logger = PluginLoggerImpl(plugin_id)

        # Create file system API
        can_read_files = PluginPermission.FILE_READ in perm_set
        can_write_files = PluginPermission.FILE_WRITE in perm_set
        fs = PluginFileSystemImpl(
            plugin_id, can_read_files, can_write_files, descriptor.path
        )

        # Create HTTP client
        can_network = PluginPermission.NETWORK_ACCESS in perm_set
        http = PluginHTTPClientImpl(plugin_id, can_network)
        self._http_clients[plugin_id] = http

        # Create database API
        can_read_db = PluginPermission.DATABASE_READ in perm_set
        can_write_db = PluginPermission.DATABASE_WRITE in perm_set
        database = PluginDatabaseImpl(plugin_id, can_read_db, can_write_db)

        # Create MCP tools API
        can_mcp = PluginPermission.MCP_TOOLS in perm_set
        mcp = PluginMCPToolsImpl(plugin_id, can_mcp)

        # Create notifications API
        can_notify = PluginPermission.NOTIFICATIONS in perm_set
        notifications = PluginNotificationsImpl(plugin_id, can_notify, self._event_bus)

        # Create communication API
        communication = PluginCommunicationImpl(plugin_id, self._event_bus)

        # Create storage API
        storage = PluginStorageImpl(plugin_id)
        self._storages[plugin_id] = storage

        # Attach API implementations to context
        # Using object.__setattr__ to bypass Pydantic validation
        object.__setattr__(context, "logger", logger)
        object.__setattr__(context, "fs", fs)
        object.__setattr__(context, "http", http)
        object.__setattr__(context, "database", database)
        object.__setattr__(context, "mcp", mcp)
        object.__setattr__(context, "notifications", notifications)
        object.__setattr__(context, "communication", communication)
        object.__setattr__(context, "storage", storage)
        object.__setattr__(context, "events", self._event_bus)

        self._contexts[plugin_id] = context
        return context

    async def destroy_context(self, plugin_id: str) -> None:
        """
        Destroy an API context.

        Args:
            plugin_id: Plugin identifier.
        """
        context = self._contexts.get(plugin_id)
        if not context:
            return

        # Clean up event listeners
        self._event_bus.remove_all(f"plugin.{plugin_id}.")

        # Close HTTP client
        http_client = self._http_clients.get(plugin_id)
        if http_client:
            await http_client.close()
            del self._http_clients[plugin_id]

        # Clear storage
        if plugin_id in self._storages:
            del self._storages[plugin_id]

        # Remove context
        del self._contexts[plugin_id]

    def emit_global_event(self, event: str, data: Any) -> None:
        """
        Emit a global event.

        Args:
            event: Event name.
            data: Event data.
        """
        self._event_bus.emit(event, data)

    def on_global_event(self, event: str, handler: Callable[..., Any]) -> None:
        """
        Subscribe to a global event.

        Args:
            event: Event name.
            handler: Event handler.
        """
        self._event_bus.on(event, handler)

    async def dispose(self) -> None:
        """Dispose of the API provider and all contexts."""
        # Close all HTTP clients
        for http_client in list(self._http_clients.values()):
            await http_client.close()
        self._http_clients.clear()

        # Clear all contexts
        self._contexts.clear()

        # Clear all storages
        self._storages.clear()

        # Clear event handlers
        self._event_bus._handlers.clear()
